mk/spksrc.cross-rust.mk

```diff
@@ -0,0 +1,249 @@
+# may be not supported for cargo
+RUST_TARGET=armv5te-unknown-linux-gnueabi
+endif
+ifeq ($(findstring $(ARCH), $(ARM7_ARCHES)),$(ARCH))
+RUST_TARGET=armv7-unknown-linux-gnueabihf
+endif
+ifeq ($(findstring $(ARCH), $(ARM8_ARCHES)),$(ARCH))
+RUST_TARGET=aarch64-unknown-linux-gnu
+endif
+ifeq ($(findstring $(ARCH), $(PPC_ARCHES)),$(ARCH))
+RUST_TARGET=powerpc-unknown-linux-gnu
+endif
+ifeq ($(RUST_TARGET),)
+$(error Arch $(ARCH) not supported)
+endif
+
+
+# Default use release channel 'stable'
+ifeq ($(strip $(RUST_RELEASE_CHANNEL)),)
+RUST_RELEASE_CHANNEL = stable
+endif
+
+install_rust_toolchain:
+	@echo "  ==> install rust toolchain [$(RUST_RELEASE_CHANNEL)]" ; \
+	rustup toolchain install $(RUST_RELEASE_CHANNEL) ;  \
+	rustup default $(RUST_RELEASE_CHANNEL) ;
+
+
+# Install rust target on demand:
+INSTALLED_RUST_TARGETS := $(shell rustup target list --installed)
+ifneq ($(findstring $(RUST_TARGET),${INSTALLED_RUST_TARGETS}),$(RUST_TARGET))
+install_rust_target:
+	@echo "  ==> install rust target [$(RUST_TARGET)]" ; \
+	rustup target install $(RUST_TARGET) ;
+
+else
+install_rust_target:
+	@echo "  ==> rust target [$(RUST_TARGET)] is already installed" ;
+
+endif
+
+
+# Set default RUST_SRC_DIR
+ifeq ($(strip $(RUST_SRC_DIR)),)
+RUST_SRC_DIR = $(WORK_DIR)/$(PKG_DIR)
+endif
+
+# Set linker environment variable
+RUST_LINKER_ENV=CARGO_TARGET_$(shell echo $(RUST_TARGET) | tr - _ | tr a-z A-Z)_LINKER
+CARGO_ENV=$(RUST_LINKER_ENV)=$(TC_PATH)$(TC_PREFIX)gcc
+
+# Use distrib folder as cargo download cache
+ENV += CARGO_HOME=/spksrc/distrib/cargo
+
+# Set the cargo parameters
+CARCO_BUILD_ARGS += --target=$(RUST_TARGET)
+CARCO_BUILD_ARGS += --path $(RUST_SRC_DIR)
+CARCO_BUILD_ARGS += --root $(STAGING_INSTALL_PREFIX)
+
+
+ifeq ($(strip $(INSTALL_TARGET)),)
+INSTALL_TARGET = rust_build_and_install_target
+endif
+
+# Default rust build and installation with cargo
+rust_build_and_install_target:
+	@echo "  ==> Cargo install rust package $(PKG_NAME)"
+	$(CARGO_ENV) cargo install $(CARCO_BUILD_ARGS)
+
+
+#####
+
+ifneq ($(REQ_KERNEL),)
+  @$(error rust modules cannot build when REQ_KERNEL is set)
+endif
+
+# Check if package supports ARCH
+ifneq ($(UNSUPPORTED_ARCHS),)
+  ifneq (,$(findstring $(ARCH),$(UNSUPPORTED_ARCHS)))
+    @$(error Arch '$(ARCH)' is not a supported architecture )
+  endif
+endif
+
+# Check minimum DSM requirements of package
+ifneq ($(REQUIRED_DSM),)
+  ifeq (,$(findstring $(ARCH),$(SRM_ARCHS)))
+    ifneq ($(REQUIRED_DSM),$(firstword $(sort $(TCVERSION) $(REQUIRED_DSM))))
+      @$(error DSM Toolchain $(TCVERSION) is lower than required version in Makefile $(REQUIRED_DSM))
+    endif
+  endif
+endif
+# Check minimum SRM requirements of package
+ifneq ($(REQUIRED_SRM),)
+  ifeq ($(ARCH),$(findstring $(ARCH),$(SRM_ARCHS)))
+    ifneq ($(REQUIRED_SRM),$(firstword $(sort $(TCVERSION) $(REQUIRED_SRM))))
+      @$(error SRM Toolchain $(TCVERSION) is lower than required version in Makefile $(REQUIRED_SRM))
+    endif
+  endif
+endif
```

Comment:
Would it be possible to move these checks into a generic Makefile to include, to avoid copy-paste.